### ISSUE TITLE
Building a Potential-like custom random variable

### ISSUE THREAD

**Issue by jamestwebber** [blei-lab/edward#443]
I have a multi-task logistic regression problem. I built a model with pymc3 and wanted to try out edward as a comparison, but in pymc3 I was using a Potential to represent a constraint and I'm having trouble translating that to this framework.

I have multiple different outputs for each sample. Each output is a logistic regression, and the outputs are linked by the fact that they all share the same feature weights&mdash;only the intercepts are different. I expect the intercepts to be monotonically increasing across the set of outputs.

Code to generate some data:
```python
import numpy as np
from scipy.special import expit
import tensorflow as tf
import edward as ed
from edward.models import Normal, Empirical, Bernoulli

n_samples = 30
n_features = 10
n_outputs = 8

x_vals = np.random.normal(size=(n_samples, n_features))
true_coef = np.random.normal(size=n_features)

xb = np.dot(x_vals, true_coef)

y_lo, y_hi = np.percentile(xb, (5., 95.))
true_ints = np.linspace(y_lo, y_hi, n_outputs) # monotonically increasing intercepts
true_y = expit(xb[...,None] + true_ints)

# MODEL
X = tf.placeholder(tf.float32, [n_samples, n_features])
beta = Normal(mu=tf.zeros(n_features), sigma=tf.ones(n_features))
intercepts = Normal(mu=tf.zeros(n_outputs), sigma=tf.ones(n_outputs))
y = Bernoulli(logits=tf.add(tf.expand_dims(ed.dot(X, beta), 1), intercepts))

# INFERENCE
T = 4000
qintercepts = Empirical(params=tf.Variable(tf.random_normal([T, n_outputs])))
qbeta = Empirical(params=tf.Variable(tf.random_normal([T, n_features])))

inference = ed.SGHMC({beta: qbeta, intercepts: qintercepts}, 
                      data={X: x_vals, y: true_y})
inference.run(step_size=1e-3)
```

The toy model works well but as the problem expands it can start to behave poorly. What I want to do is add a likelihood penalty when the intercept vector is not monotonically increasing. previously, I'd count how many mis-ordered intercepts I had, and square the sum:

```python
# in python
-np.triu(intercepts[:,None] >= intercepts[None,:], 1).sum()**2
# for edward (I think)
-(tf.reduce_sum(tf.matrix_band_part(tf.expand_dims(intercepts, 1) 
                                    >= tf.expand_dims(intercepts, 0), 1, -1))**2)
```

But I'm not sure how to incorporate that as a non-sampled custom variable in the model (and @dustinvtran suggested I make an issue to discuss).

This particular model probably isn't of general interest, but I'd be happy to write up an example to demonstrate how to do this.

**Comment by dustinvtran** [comment]
interesting. this doesn't seem tied to the likelihood but the prior. are you saying you'd like this penalty term to be added to all log joint density calls?

**Comment by jamestwebber** [comment]
Yeah, that's right. The [pymc documentation](https://pymc-devs.github.io/pymc/modelbuilding.html#the-potential-class) explains the concept better than I can. You're just calculating some function of the existing variables as a way to add a specific penalty to the probability density.

In edward, perhaps all this means is that I create a custom random variable with the desired `_log_prob` method and a `_sample_n` method that calculates the value deterministically from the parent nodes rather than generating a random sample.

**Comment by jamestwebber** [comment]
I am having a lot of trouble figuring out how to get this working. It's hard to parse all the interactions between the tensorflow library and edward and figure out what needs to be written (and how) to make it compatible with sampling. Any guidance would be much appreciated.

**Comment by jamestwebber** [comment]
For this example it looks like the solution is pretty simple: the potential is just a function of the `intercepts` variable so it can be introduced by modifying the `_log_prob` method there:

```python
class IntPotential(Normal):
    def _log_prob(self, value):
        bool_m = tf.expand_dims(value, 1) >= tf.expand_dims(value, 0)
        triu_m = tf.matrix_band_part(tf.to_float(bool_m), 0, -1)
        
        return super(IntPotential, self)._log_prob(value) - tf.pow(tf.reduce_sum(triu_m) - value.get_shape()[0].value, 2)

intercepts = IntPotential(mu=tf.zeros(n_conditions), sigma=tf.ones(n_conditions))
X = tf.placeholder(tf.float32, [n_samples, n_features])
beta = Normal(mu=tf.zeros(n_features), sigma=tf.ones(n_features))
y = Bernoulli(logits=tf.add(tf.expand_dims(ed.dot(X, beta), 1), intercepts))
```

That appears to be working here. But for a more general solution (i.e. a potential that incorporated multiple nodes) I don't think it would.

**Comment by dustinvtran** [comment]
pasting chat from Gitter:

> James Webber @jamestwebber Feb 10 16:40
@dustinvtran I think I got it to work, at least in this case. because my potential is a deterministic function of a single variable, I just needed to subclass that RV and write a new _log_prob method for it
but in the general case i have no idea how to do it...seems like it shouldn't be that hard to implement a node that doesn't get sampled, but I didn't figure it out

> Dustin Tran @dustinvtran Feb 10 18:17
nice!
in the general case it sounds difficult but is a very interesting design problem
we can always force it to be at the likelihood level, and raise errors if another random variable uses it as a prior. and we can raise errors for algorithms that assume one can sample from the likelihood.
for example, one can't use GANInference for this model class, but one can use HMC or KLqp.
